### DISCUSSION TITLE
ownership issue

### DISCUSSION THREAD

**Discussion by schmurfy** [tauri-apps/tauri#7435] [General]
I think that's more a problem related to rust than tauri specifically but I failed to reproduce it without tauri so I will try asking here first:

I am trying to implement a managed state to share the client once it is initialized, I reduced the code to be more readable but the `client` method would create the client if client contains `None` and then return it.

```rust
use kube::Client;
use crate::kube::get_client;

pub struct AppState {
  client: Option<Client>,
}

impl AppState {
  pub fn client3(&self) -> Option<Client> {
    match &self.client {
      Some(c) => Some(c.clone()),
      None => None,
    }
  }

  pub fn client2(&mut self) -> Option<Client> {
    match &self.client {
      Some(c) => Some(c.clone()),
      None => None,
    }
  }
}
``` 

My issue is there in my commands:
```rust
#[command]
pub async fn watch_pods(app_state: tauri::State<'_, AppState>) -> Result<(), MyError> {
  let client = app_state.client2().unwrap();
  crate::kube::pod::watch(client, |p| {
    // [...]
  }).await.unwrap();
  Ok(())
}
```
```rust
// watch function signature for reference
pub async fn watch<F: Fn(PodEvent)>(client: Client, f: F) -> Result<(), kube::Error> {
  // [...]
}
```

If I call `client3` it works fine, no compiler error but if I use `client2` I get this error:
```
error[E0596]: cannot borrow data in dereference of `tauri::State<'_, AppState>` as mutable
  --> src/commands.rs:47:16
   |
47 |   let client = app_state.client2().unwrap();
   |                ^^^^^^^^^^^^^^^^^^^ cannot borrow as mutable
   |
   = help: trait `DerefMut` is required to modify through a dereference, but it is not implemented for `tauri::State<'_, AppState>`
```

Despite reading again and again everything I can find in the rust book and on the net about how ownership works I cannot understand the error and how to solve it.

As far as I understand the `Client` is owned by the Option which is owned by my state struct but when I clone it to return it the ownership of that new object moves to the `watch_pods` function and those assumptions seems correct but all hell break loose when I unwrap the Option.

When using `client3` with an immutable reference it works but not with `client2` using a mutable reference and that's where I am lost, I don't see what link the clone Option to `self` and how the type of reference I had on it when calling clone matters.

Any help is welcome, I have been stuck on this for two days now and really want to understand.

**Comment by FabianLars** [comment]
The Option doesn't actually matter here, it fails before that already. In your command you accept an immutable instance of `AppState` but the client2 function needs it to be mutable. And i think this is not an issue for you without tauri because tauri::State is basically a fancy reference and not an owned value (it works similar to std::sync::Arc if you wanna look into it).

**Reply by schmurfy** [reply]
Thank you, that makes more sense.
If the state is an immutable reference, how can I update it ?

I have looked at `examples/state` but I admit it confused me more than anything, I guess I am learning a bit too much at the same time xD

How would you do what I try to achieve here ?(initialize the client on first call and then reuse it)

**Reply by FabianLars** [reply]
If you need a mutable reference then i think you won't get around interior mutability via a [Mutex](https://doc.rust-lang.org/std/sync/struct.Mutex.html) for example.
